File: 1-D Dynamic Programming/91 Decode Ways.go
package main

import (
	"fmt"
	"strconv"
)

func main() {
	fmt.Println(numDecodings("11106"))
}

func numDecodings(s string) int {
	seen := map[int]int{}
	var dp func(int) int
	dp = func(i int) int {
		fmt.Println(i)
		if i >= len(s) {
			return 1
		}
		if string(s[i]) == "0" {
			return 0
		}
		if val, ok := seen[i]; ok {
			return val
		}

		if i < len(s)-1 {
			if num, _ := strconv.Atoi(s[i : i+2]); num <= 26 {
				seen[i] = dp(i+1) + dp(i+2)
				return seen[i]
			}
		}
		seen[i] = dp(i + 1)
		return seen[i]
	}
	return dp(0)
}

File: 1-D Dynamic Programming/91 Decode Ways.py
from functools import cache

class Solution:
    def numDecodings(self, s: str) -> int:
        seen = {} # index : no. solutions
        @cache
        def dp(i: int) -> int:
            # base cases
            if i >= len(s):
                return 1
            if s[i] == "0":
                return 0
            if i in seen:
                return seen[i]

            if int(s[i:i+2]) <= 26 and i < len(s) - 1:
                seen[i] =  dp(i+1) + dp(i+2)
                return seen[i]
            seen[i] = dp(i+1)
            return seen[i]
        return dp(0)

sol = Solution()
print(sol.numDecodings("1111111111111111111111111111"))